File: disjoint_set/detect_cycle_using_dsu.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends

class Solution {
  int root(int i, vector<int>& p) {
    while (i != p[i]) {
      p[i] = p[p[i]];
      i = p[i];
    }
    return i;
  }
  bool connected(int i, int j, vector<int>& p) {
    int r1 = root(i, p);
    int r2 = root(j, p);
    return r1 == r2;
  }
  void connect(int i, int j, vector<int>& p) {
    int r1 = root(i, p);
    int r2 = root(j, p);
    p[r1] = r2;
  }

 public:
  // Function to detect cycle using DSU in an undirected graph.
  int detectCycle(int V, vector<int> adj[]) {
    // Code here
    vector<int> p(V);
    for (int i = 0; i < V; ++i) {
      p[i] = i;
    }
    for (int i = 0; i < V; ++i) {
      for (int j = 0; j < adj[i].size(); ++j) {
        // cout << i << ' ' << adj[i][j] << endl;
        if (adj[i][j] < i) continue;
        if (connected(i, adj[i][j], p)) {
          return 1;
        } else {
          connect(i, adj[i][j], p);
        }
      }
    }
    return 0;
  }
};

//{ Driver Code Starts.
int main() {
  int tc;
  cin >> tc;
  while (tc--) {
    int V, E;
    cin >> V >> E;
    vector<int> adj[V];
    for (int i = 0; i < E; i++) {
      int u, v;
      cin >> u >> v;
      adj[u].push_back(v);
      adj[v].push_back(u);
    }
    Solution obj;
    int ans = obj.detectCycle(V, adj);
    cout << ans << "\n";
  }
  return 0;
}
// } Driver Code Ends

File: disjoint_set/detect_cycle_using_dsu.py
class Solution:
    # Function to detect cycle using DSU in an undirected graph.
    def detectCycle(self, V, edge):
        # Code here
        edge.sort()
        id = list(c for c in range(V))
        count = [1] * V

        def root(u):
            while u != id[u]:
                id[u] = id[id[u]]
                u = id[u]
            return u

        def merge(u, v):
            pu = root(u)
            pv = root(v)
            if count[pu] < count[pv]:
                id[pu] = pv
                count[pv] += count[pu]
            else:
                id[pv] = pu
                count[pu] += count[pv]

        for i in range(len(edge)):
            if i != 0 and edge[i] == edge[i - 1]:
                continue
            if root(edge[i][0]) == root(edge[i][1]):
                return 1
            merge(edge[i][0], edge[i][1])

        return 0


if __name__ == "__main__":
    T = int(input())
    for i in range(T):
        V, E = map(int, input().split())
        edge = []
        for _ in range(E):
            u, v = map(int, input().split())
            edge.append((min(u, v), max(u, v)))

        if E > V - 1:
            print(1)
            continue
        obj = Solution()
        ans = obj.detectCycle(V, edge)
        print(ans)
    exit(0)
